src/logging/timer.h

```diff
@@ -60,24 +59,21 @@ class Timer
 
     std::string LogMsg(const std::string& msg)
     {
-        const auto end_time = GetTime<std::chrono::microseconds>() - m_start_t;
-        if (m_start_t.count() <= 0) {
+        const auto dur{std::chrono::steady_clock::now() - m_start_t};
+        if (m_start_t == decltype(m_start_t){}) {
             return strprintf("%s: %s", m_prefix, msg);
         }
-
-        if constexpr (std::is_same<TimeType, std::chrono::microseconds>::value) {
-            return strprintf("%s: %s (%iμs)", m_prefix, msg, end_time.count());
-        } else if constexpr (std::is_same<TimeType, std::chrono::milliseconds>::value) {
-            return strprintf("%s: %s (%.2fms)", m_prefix, msg, end_time.count() * 0.001);
-        } else if constexpr (std::is_same<TimeType, std::chrono::seconds>::value) {
-            return strprintf("%s: %s (%.2fs)", m_prefix, msg, end_time.count() * 0.000001);
-        } else {
-            static_assert(ALWAYS_FALSE<TimeType>, "Error: unexpected time type");
+        if (dur > 1s) {
+            return strprintf("%s: %s (%.2fs)", m_prefix, msg, CountSecondsDouble(dur));
+        }
+        if (dur > 1ms) {
+            return strprintf("%s: %s (%.2fms)", m_prefix, msg, CountMillisecondsDouble(dur));
         }
+        return strprintf("%s: %s (%.2fμs)", m_prefix, msg, CountMicrosecondsDouble(dur));
```

Comment:
Maybe inverse the order of the duration checks from the smallest unit to the largest.

Reply:
Can you explain why?

Reply:
In higher frequency timing logs like lock, the smaller units seem to be more frequent. Debug build on a slow system:

```
2022-04-19T17:59:03Z [loadblk] Enter: lock contention cs_main, validation.cpp:4672 started
2022-04-19T17:59:03Z [loadblk] Enter: lock contention cs_main, validation.cpp:4672 completed (253.02μs)
2022-04-19T17:59:03Z [msghand] Enter: lock contention cs_main, net_processing.cpp:4196 started
2022-04-19T17:59:03Z [msghand] Enter: lock contention cs_main, net_processing.cpp:4196 completed (14.28ms)
2022-04-19T17:59:03Z [loadblk] Enter: lock contention cs_main, validation.cpp:4672 started
2022-04-19T17:59:03Z [scheduler] Enter: lock contention ::cs_main, validation.cpp:5151 started
2022-04-19T17:59:03Z [loadblk] Enter: lock contention cs_main, validation.cpp:4672 completed (157.08ms)
2022-04-19T17:59:03Z [msghand] Enter: lock contention ::cs_main, validation.cpp:5151 started
2022-04-19T17:59:03Z [scheduler] Enter: lock contention ::cs_main, validation.cpp:5151 completed (24.14ms)
2022-04-19T17:59:03Z [loadblk] Enter: lock contention cs_main, validation.cpp:4672 started
2022-04-19T17:59:03Z [loadblk] Enter: lock contention cs_main, validation.cpp:4672 completed (10.48μs)
2022-04-19T17:59:03Z [msghand] Enter: lock contention ::cs_main, validation.cpp:5151 completed (7.92ms)
2022-04-19T17:59:03Z [loadblk] Enter: lock contention cs_main, validation.cpp:4672 started
2022-04-19T17:59:03Z [loadblk] Enter: lock contention cs_main, validation.cpp:4672 completed (5.57μs)
2022-04-19T17:59:03Z [msghand] Enter: lock contention ::cs_main, validation.cpp:5151 started
2022-04-19T17:59:03Z [msghand] Enter: lock contention ::cs_main, validation.cpp:5151 completed (85.71μs)
2022-04-19T17:59:03Z [loadblk] Enter: lock contention cs_main, validation.cpp:4672 started
2022-04-19T17:59:03Z [loadblk] Enter: lock contention cs_main, validation.cpp:4672 completed (4.68μs)
2022-04-19T17:59:03Z [msghand] Enter: lock contention m_mutex, random.cpp:416 started
2022-04-19T17:59:03Z [msghand] Enter: lock contention m_mutex, random.cpp:416 completed (62.23μs)
2022-04-19T17:59:03Z [msghand] Enter: lock contention cs_main, net_processing.cpp:4668 started
2022-04-19T17:59:03Z [msghand] Enter: lock contention cs_main, net_processing.cpp:4668 completed (911.14μs)
```

Reply:
I mean the order is intentional and it can not be changed without changing the behaviour of the code.

Reply:
I mean, you could invert both the logic and the order?

```c++
if (dur <= 1ms) { /* microseconds */ }
else if (dur <= 1s) { /* milliseconds */ }
else { /* seconds */ }
```

Doing one integer comparison instead of two prior to a printf seems like an amazing level of microoptimisation though...

Reply:
It's new code anyway and low-hanging, so mentioned it.

Reply:
And the existing code checks from small to large.

Reply:
How about changing
```c++
strprintf("%s: %s (%.2fμs)", m_prefix, msg, CountMicrosecondsDouble(dur))
```

to

```c++
strprintf("%s: %s (%.5fms)", m_prefix, msg, CountMillisecondsDouble(dur))
``` 

? Then you'd see "10.23s", "10.23ms" and "0.01023ms" which might be easier to follow?

Reply:
There aren't any in the sample above, but it's the big ones in seconds (s) that I find harder to pick out when eyeballing the mix.  The μs ones stand out pretty well to my eye. Sample with seconds:
```
2022-04-19T17:58:03Z [net] Enter: lock contention m_nodes_mutex, net.cpp:1294 started
2022-04-19T17:58:03Z [net] Enter: lock contention m_nodes_mutex, net.cpp:1294 completed (4.16μs)
2022-04-19T17:58:03Z [opencon] Enter: lock contention m_nodes_mutex, net.cpp:366 started
2022-04-19T17:58:03Z [opencon] Enter: lock contention m_nodes_mutex, net.cpp:366 completed (4.43μs)
2022-04-19T17:58:03Z [net] Enter: lock contention m_nodes_mutex, net.cpp:1294 started
2022-04-19T17:58:03Z [net] Enter: lock contention m_nodes_mutex, net.cpp:1294 completed (4.28μs)
2022-04-19T17:58:03Z [opencon] Enter: lock contention cs_main, net_processing.cpp:1229 started
2022-04-19T17:58:05Z [msghand] Enter: lock contention ::cs_main, validation.cpp:5151 started
2022-04-19T17:58:05Z [loadblk] Enter: lock contention cs_main, validation.cpp:4672 completed (12.17s)
2022-04-19T17:58:05Z [opencon] Enter: lock contention cs_main, net_processing.cpp:1229 completed (2.23s)
2022-04-19T17:58:05Z [msghand] Enter: lock contention ::cs_main, validation.cpp:5151 completed (4.40ms)
2022-04-19T17:58:05Z [scheduler] Enter: lock contention ::cs_main, validation.cpp:5151 completed (11.81s)
2022-04-19T17:58:05Z [msghand] Enter: lock contention cs_main, validation.cpp:2906 started
2022-04-19T17:58:05Z [loadblk] Enter: lock contention cs_main, validation.cpp:4672 started
2022-04-19T17:58:05Z [msghand] Enter: lock contention cs_main, validation.cpp:2906 completed (558.78μs)
2022-04-19T17:58:05Z [loadblk] Enter: lock contention cs_main, validation.cpp:4672 completed (735.17μs)
```

Reply:
Sorry, tangential, but: wouldn't it be better to log "foo.cpp:123 held the cs_main lock for 20 seconds" instead of "bar.cpp:456 was waiting for the cs_main lock for 20 seconds" ? In theory, you could then only report when some bit of code held a lock for an "unusually" long amount of time, and could do so even if no other thread happened to also want it... I just don't really see the point of a log entry for "m_nodes_mutex net.cpp:1294 (4.28μs)" -- that doesn't really seem like actionable information.

Reply:
I think it's valuable information to show who waited, because just holding a lock when nobody else needs it isn't an issue. But what I'd really like to see is who actually blocked someone else.

I'd personally prefer exact timestamps from e.g. `std::chrono::steady_clock` in the best precision possible in the logs, then one could extract a lot of useful information from it. E.g. who actually blocked whom and for how long. It would then also be possible to generate input for e.g. [chrome tracing](https://slack.engineering/chrome-tracing-for-fun-and-profit/) which is a simple JSON format.

Reply:
I disagree. If something took a short time, why would anyone care if it took 900 nanoseconds or 1700 nanoseconds? After all this isn't a microbenchmark and real world fluctuations are larger than nanosecond precision. Also, nanosecond precision didn't exist before, so introducing it can be done in a separate pull or a follow-up.

Changing the content of the lock contention logging or chrome tracing also appear largely unrelated to a change to the formatting of durations in logs.